### PR TITLE
Add py310 support

### PR DESCRIPTION
**Pull request by ocefpaf**
Added a test matrix to be sure we find some corner cases.

Closes https://github.com/ericdill/depfinder/issues/69

- [X] add py310 support
- [X] make `stdlib-list` a py<310 dependency
- [X] fix false positive in our tests

### CODE REVIEW

File: depfinder/inspection.py
Comment (zmoon) [inline depfinder/inspection.py:51]:
```suggestion
if sys.version_info < (3, 10):
```
Wouldn't this be fine? And then `packaging` not needed.

Comment (ocefpaf) [reply depfinder/inspection.py]: This is still a draft and pretty much in flux, don't worry about it now. BTW,  I already removed that b/c I refactor that part.

PS: I'd love a review when I finished it though :grimacing:

Comment (zmoon) [reply depfinder/inspection.py]: Ok!

File: .github/workflows/tests.yml
Comment (zmoon) [inline .github/workflows/tests.yml:31]: This can go in the job defaults or top-level defaults to simply things.

https://github.com/mamba-org/provision-with-micromamba#important

File: .github/workflows/tests.yml
Comment (zmoon) [inline .github/workflows/tests.yml:26]:
```suggestion
        uses: mamba-org/provision-with-micromamba@v12
```

File: .github/workflows/tests.yml
Comment (zmoon) [inline .github/workflows/tests.yml:28]: [Seems](https://github.com/mamba-org/provision-with-micromamba#environment-name) that you should add `environment-name` in the case of `environment-file: false`?

File: .github/workflows/tests.yml
Comment (zmoon) [inline .github/workflows/tests.yml:27]: I was thinking about the caching options but I guess that wouldn't help since no conda env yml file

File: .github/workflows/tests.yml
Comment (zmoon) [inline .github/workflows/tests.yml:36]:
```suggestion

```

File: .github/workflows/tests.yml
Comment (zmoon) [inline .github/workflows/tests.yml:33]: Could add `python` and `pip` in [extra-specs](https://github.com/mamba-org/provision-with-micromamba#extra-specs) above and then just update the environment here instead of create. And then maybe the activate lines wouldn't be needed?

File: depfinder/inspection.py
Comment (zmoon) [inline depfinder/inspection.py:39]: Could just be me but I find `stdliblist` a bit ugly / hard to read, maybe the module could be `stdlib_list_` instead or something

**Comment by zmoon** [comment]
Also did you see the
```
test.py::test_for_smoke
  <unknown>:485: DeprecationWarning: invalid escape sequence \s
```
in the test run? Not part of this PR though.

File: depfinder/inspection.py
Comment (ocefpaf) [reply depfinder/inspection.py:39]: Super ugly! I'm terrible with names. I'm OK with `stdlib_list_`. Let's see what @ericdill thinks and we'll rename.

File: .github/workflows/tests.yml
Comment (ocefpaf) [reply .github/workflows/tests.yml:33]: I find that a bit confusing b/c I like the CIs to be as close as what a user would do locally.

File: .github/workflows/tests.yml
Comment (ocefpaf) [reply .github/workflows/tests.yml:27]: Probably. I only use cache for big envs, this one is not that heavy but we can try that in another PR I already added too much in this one.

File: .github/workflows/tests.yml
Comment (ocefpaf) [reply .github/workflows/tests.yml:28]: That is optional: https://github.com/mamba-org/provision-with-micromamba/issues/54

File: depfinder/stdliblist.py
Comment (zmoon) [inline depfinder/stdliblist.py:10]: This *should* be the case now, but if stdlib-list was somehow missing in Python <=3.9, then error message could be a bit confusing?

**Comment by ocefpaf** [comment]
> in the test run? Not part of this PR though.

Not sure where that is coming from but we should probably fix it in another PR.

File: .github/workflows/tests.yml
Comment (zmoon) [reply .github/workflows/tests.yml:33]: That is understandable. Why not just use setup-python instead then? I think provision-with-micromamba is great, but if you don't even have a conda env, then maybe it is overkill.

File: .github/workflows/tests.yml
Comment (ocefpaf) [reply .github/workflows/tests.yml]: Done!

File: .github/workflows/tests.yml
Comment (ocefpaf) [reply .github/workflows/tests.yml:30]: Indeed this is safer but I have to confess I've been "riding the wave" until today. Let's pin it!

PS: b/c GH releases can be modified to only way to be really safe is to use a specific commit.

File: .github/workflows/tests.yml
Comment (ocefpaf) [reply .github/workflows/tests.yml:36]: For this project it would not be that different, you are correct. But I only use conda packages so this makes it easier for me to port stuff from one project to another.

File: depfinder/stdliblist.py
Comment (ocefpaf) [reply depfinder/stdliblist.py:10]: It is listed as a dependency for py<3.10, so the only way that scenario would happen is in a broken installation with "no-deps"  and people doing that are usually experts. The error message would be:

```
AttributeError: module 'sys' has no attribute 'stdlib_module_names'
```

That is, IMO, the correct information an advanced users should know but I accept suggestions for a better one.

File: .github/workflows/tests.yml
Comment (zmoon) [reply .github/workflows/tests.yml:36]: That's fair!

File: .github/workflows/tests.yml
Comment (zmoon) [reply .github/workflows/tests.yml:32]: Thanks for sharing that

File: depfinder/stdliblist.py
Comment (zmoon) [reply depfinder/stdliblist.py:10]: Yeah, I suppose you are right, that makes sense.

I guess I would lean towards switching the `try` and `except` parts or doing a `sys.version_info` check.

And I know you're going another way here, but I also feel like it could be useful to have this in stdlib-list, e.g. to support listing the 3.10 stdlib from an older Python version, and to reduce the number of times this code here would have to be written perhaps.

File: depfinder/inspection.py
Comment (ocefpaf) [reply depfinder/inspection.py:39]: Actually, let's hold this one for a bit. I forgot I'm also a maintainer of stdlib-list and these changes should be there instead.

File: depfinder/stdliblist.py
Comment (ocefpaf) [reply depfinder/stdliblist.py:10]: Yep. It makes sense to put in stliblib-list. I completely forgot I am a maintainer there and I can port this over tomorrow.

File: depfinder/inspection.py
Comment (ericdill) [reply depfinder/inspection.py:39]: > Let's see what @ericdill thinks and we'll rename.

FWIW you don't need my approval for anything in this project. do whatever you think is best. i trust you on this stuff

**Review by beckermr — APPROVED** [review]

**Comment by beckermr** [comment]
My approval never was or is needed here, but I am LGTMing this in an effort to get it merged ASAP. 

Thank you @ocefpaf!

**Comment by ericdill** [comment]
i pushed a change to fix the tests but now we're failing because of the change introduced in #76. the reason for this is that python-stdlib-list has the builtin as "concurrent.futures". so when we're using python-stdlib-list, that's the thing that gets reported. The new way we're finding modules,
```
    builtin_modules = list(set(list(sys.stdlib_module_names) + list(sys.builtin_module_names)))
```

only has the top-level namespace (i.e., it only shows `concurrent`).

so now i think the solution probably has to be to have separate tests for py<=3.9 and py>=3.10. i'll give a think on how to implement that

**Comment by ericdill** [comment]
🤞🏻 [584c8a9](https://github.com/ericdill/depfinder/pull/74/commits/584c8a9a6ca2c74f3390a98d9e78ca16aa349249) 🤞🏻 should fix it

File: depfinder/inspection.py
Comment (ericdill) [reply depfinder/inspection.py:39]: i'd prefix this to be `_stdlib_list` but it's an internal detail so we can fix it in post

**Comment by ericdill** [comment]
>   <unknown>:485: DeprecationWarning: invalid escape sequence \s

@zmoon iirc this was from some weird windows encoding thing that CJ hit a while back. it's in one of the test files somewhere i think. could be a thing to look into later

**Comment by ericdill** [comment]
yolo

**Comment by ericdill** [comment]
@ocefpaf it's embarassing to admit but i have no idea how to release a new version of this

**Comment by ericdill** [comment]
oh actually looks like @beckermr did the last release. want to do another?

**Comment by ocefpaf** [comment]
> @ocefpaf it's embarassing to admit but i have no idea how to release a new version of this

I'm just tagging and using GitHub auto generated changelog. See https://github.com/ericdill/depfinder/releases/tag/V2.9.0

I'll add some extra stuff to automatically push to PyPI soon but I can push this one manually.

**Comment by ocefpaf** [comment]
@ericdill I do plan to move most of this stuff to `stdlib-list` at some point and "archive" that project. In theory we won't need it again for Python >=3.10.

**Comment by ericdill** [comment]
thanks for releasing!

conda-forge PR https://github.com/conda-forge/depfinder-feedstock/pull/29

**Comment by beckermr** [comment]
Something is failing for python 3.10 in the feedstock pr. Can you all look?

**Pull request merged**
